language/thingml/src/org/thingml/xtext/validation/checks/FunctionUsage.xtend

```diff
@@ -1,281 +1,281 @@
-					val msg = "Function "+function.name+" is called with an erroneous parameter. Expected "+Types.toString(expected)+", called with "+Types.toString(TyperHelper.getBroadType(actual));
-					if (parent instanceof EList)
-						error(msg, o.eContainer, o.eContainingFeature, (parent as EList).indexOf(o) ,"function-call-wrong-parameter-type")
-					else
-						error(msg, o.eContainer, o.eContainingFeature ,"function-call-wrong-parameter-type")
-				}
-			}
-		]
-	}
-}
+package org.thingml.xtext.validation.checks
+
+import java.util.List
+import org.eclipse.emf.common.util.EList
+import org.eclipse.emf.ecore.EObject
+import org.eclipse.xtext.validation.Check
+import org.thingml.xtext.constraints.ThingMLHelpers
+import org.thingml.xtext.constraints.Types
+import org.thingml.xtext.helpers.ActionHelper
+import org.thingml.xtext.helpers.AnnotatedElementHelper
+import org.thingml.xtext.helpers.FunctionWithMultipleImplem
+import org.thingml.xtext.helpers.ThingHelper
+import org.thingml.xtext.helpers.TyperHelper
+import org.thingml.xtext.helpers.UnimplementedFunction
+import org.thingml.xtext.thingML.Action
+import org.thingml.xtext.thingML.ActionBlock
+import org.thingml.xtext.thingML.ConditionalAction
+import org.thingml.xtext.thingML.Expression
+import org.thingml.xtext.thingML.ExternStatement
+import org.thingml.xtext.thingML.Function
+import org.thingml.xtext.thingML.FunctionCallExpression
+import org.thingml.xtext.thingML.FunctionCallStatement
+import org.thingml.xtext.thingML.LoopAction
+import org.thingml.xtext.thingML.PropertyReference
+import org.thingml.xtext.thingML.ReturnAction
+import org.thingml.xtext.thingML.Thing
+import org.thingml.xtext.thingML.ThingMLPackage
+import org.thingml.xtext.thingML.VariableAssignment
+import org.thingml.xtext.validation.ThingMLValidatorCheck
+import org.thingml.xtext.validation.TypeChecker
+
+class FunctionUsage extends ThingMLValidatorCheck {
+	
+	@Check(FAST)
+	def checkThread(Function f) {
+		if (AnnotatedElementHelper.isDefined(f, "fork_thread", "true") || AnnotatedElementHelper.hasFlag(f, "fork_thread")) {
+			if (f.typeRef !== null && f.typeRef.type !== null) {
+				val msg = "Function " + f.name + " forks a thread. It cannot return a result."
+				warning(msg, f, ThingMLPackage.eINSTANCE.namedElement_Name, "function-thread-cannot-return", f.name);
+			}
+		}
+	}
+	
+	@Check(NORMAL)
+	def checkUsage(Function f) {
+		val thing = ThingMLHelpers.findContainingThing(f)
+		//Checks if the containing thing calls the function
+		if (ActionHelper.getAllActions(thing, FunctionCallStatement).exists[call | call.function == f || call.function.name == f.name])
+			return;
+		if (ThingMLHelpers.getAllExpressions(thing, FunctionCallExpression).exists[call | call.function == f || call.function.name == f.name])
+			return;			
+		
+		//if not, check if any thing that includes the function actually calls it
+		val isUsedByIncludingThings = ThingHelper.allIncludingThings(thing)

```

Comment:
that should fix #306 as we now check including things (things that import the function directly or not), rather than included thing (which had no chance to import the function).